docs/content/docs/administration/options.md
+++
title = "Configuration options"
weight = 31
+++

selfoss offers the following configuration parameters. You can set the config parameters in the `config.ini` file.

### `db_type`
<div class="config-option">

database type (`sqlite`, `mysql` or `pgsql`)
</div>

### `db_file`
<div class="config-option">

location of database file for SQLite
</div>

### `db_host`
<div class="config-option">

address/hostname of the database server for MySQL/PostgreSQL
</div>

### `db_database`
<div class="config-option">

Name of the MySQL or PostgreSQL database
</div>

### `db_username`
<div class="config-option">

Database username for MySQL or PostgreSQL
</div>

### `db_password`
<div class="config-option">

Database password for MySQL or PostgreSQL.

For PostgreSQL on local server, [peer authentication](https://www.postgresql.org/docs/18/auth-peer.html) might be more secure an convenient, as it does not require a password. This is also available as [`unix_socket` authentication plugin](https://mariadb.com/docs/server/reference/plugins/authentication-plugins/authentication-plugin-unix-socket) on MariaDB.

Prefer [`db_password_file`](#db-password-file) if the configuration is readable by other users.
</div>

### `db_password_file`
<div class="config-option">

A path to a file containing database password.

This allows storing the password outside of the configuration, which may be more secure if the configuration is readable by other users. It is also allows provisioning the password separately using some secrets management system.

To avoid leaking the password to other services, the file should [only be accessible](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#23-access-control) by the user running the PHP server used by selfoss.
</div>

### `db_prefix`
<div class="config-option">

Table prefix for MySQL/SQLite databases. This is useful to avoid conflicts when you are sharing a database with another application.
</div>

### `db_port`
<div class="config-option">

Port for database connections. By default `3306` will be used for MySQL and `5432` for PostgreSQL.
</div>

### `db_socket`
<div class="config-option">

A UNIX domain socket used for connecting to the MySQL database server. Usually, you want to use `db_host=localhost`, which should use the default socket path (typically `/run/mysqld/mysqld.sock` for MySQL or `/run/postgresql` for PostgreSQL) but if you need to specify a different location, you can. This is orthogonal to `db_host` option.
</div>

### `debug`
<div class="config-option">

Level of detail for diagnostics. Enable this for debug traces if you encounter selfoss internal errors or are developing selfoss.

* `0` (default) – Debugging is disabled, errors will only be [logged](#logger-level).
* `1` – Debugging is enabled, any error or warning will terminate the application and cause debugging information to be printed into the web browser.
* `2` – Same as `1` but additional information (HTTP requests) will be logged when fetching feed content.
</div>

### `logger_destination`
<div class="config-option">

By default, the logs are saved to `data/logs/default.log` but you can choose a different file by specifying a file path prefixed by `file:`. Setting `file:php://stderr` is especially useful when running selfoss on a PaaS or inside Docker. Alternately, you can set the option to `error_log` to redirect the messages to [SAPI error log](https://secure.php.net/manual/en/function.error-log.php) – handy for PHP-FPM, which [discards stderr](https://secure.php.net/manual/en/install.fpm.configuration.php#catch-workers-output) by default.
</div>

### `logger_level`
<div class="config-option">

set logging level – following logging levels are available: `EMERGENCY`, `ALERT`, `CRITICAL`, `ERROR`, `WARNING`, `NOTICE`, `INFO`, `DEBUG`. Additionally, you can use `NONE` pseudo-level to turn the logging off completely.

Use this for troubleshooting on updating feeds (but be aware that the log file can become very large.)
</div>

### `items_perpage`
<div class="config-option">

number of entries per page on your stream
</div>

### `items_lifetime`
<div class="config-option">

Number of days since the item has been last seen after which it can be deleted. In the upcoming selfoss 2.20, you can set it to `0` to disable item deletion. Starred items will never be deleted.
</div>

### `base_url`
<div class="config-option">

base URL of the selfoss page; use this option if you use a ssl proxy which changes the `$_SERVER` globals, most notably the URL path in which the app is installed.
</div>

### `username`
<div class="config-option">

username for optional login. Just set username and password for enabling login.
</div>

### `password`
<div class="config-option">

password hash for optional login. You can generate a password hash by using following page of your selfoss installation. https://your_selfoss_url.com/password
</div>

### `password_file`
<div class="config-option">

Alternative to [`password`](#password) option, a path to a file containing the password hash. This is useful for provisioning the password using a secrets management system.
</div>

### <del>`salt`</del> (deprecated)
<div class="config-option">

salt for hashing the password (see [Wikipedia](https://en.wikipedia.org/wiki/Salt_(cryptography))). Not used for passwords generated using selfoss 2.19 or newer.
</div>

### `public`
<div class="config-option">

if you use login (`username` and `password` are set), you can allow guests to see your stream. Enter `1` for enabling this write-protected mode.
</div>

### `rss_title`
<div class="config-option">

title of the generated rss feed
</div>

### `rss_max_items`
<div class="config-option">

maximum amount of items in the generated rss feed
</div>

### `rss_mark_as_read`
<div class="config-option">

set this to `1` to automatically mark items as read after they appeared in selfoss’s RSS.
</div>

### `homepage`
<div class="config-option">

set here your preferred homepage. Choose between `newest`, `unread` and `starred`. It is also possible to configure a tag (e.g. `unread/tag-yourtag`) or a source (e.g. `newest/source-123`). Default = `newest`.
</div>

### `auto_mark_as_read`
<div class="config-option">

set this to `1` to automatically mark an item as read on opening it.
</div>

### `auto_collapse`
<div class="config-option">

set this to `1` to automatically collapse an item when another one is opened.
</div>

### `auto_stream_more`
<div class="config-option">

set this to `0` to disable automatic loading of more items when you scroll down. Click a button at the bottom of the page will be required instead.
</div>

### `open_in_background_tab`
<div class="config-option">

set this to `1` to try to make <kbd>v</kbd> shortcut open articles in new background tab. This [does not work in Chromium based browsers](https://crbug.com/431335).
</div>

### `language`
<div class="config-option">

set `0` or leave empty for auto detection (browser language) or use one of the following language codes:

* Catalan: `ca`
* Chinese (Simplified): `zh-CN`
* Chinese (Traditional): `zh-TW`
* Czech: `cs`
* Dutch: `nl`
* English: `en`
* English (United Kingdom): `en-GB`
* Estonian: `et`
* Finnish: `fi`
* French: `fr`
* French (Canada): `fr-CA`
* Galician: `gl`
* German: `de`
* Hebrew: `he`
* Hungarian: `hu`
* Italian: `it`
* Indonesian: `id`
* Japanese: `ja`
* Latvian: `lv`
* Norwegian Bokmål: `nb`
* Polish: `pl`
* Portuguese: `pt`
* Portuguese (Brazil): `pt-BR`
* Romansh: `rm`
* Russian: `ru`
* Slovak: `sk`
* Spanish: `es`
* Swedish: `sv`
* Tamil: `ta` (will only be available in selfoss 2.20 or later)
* Turkish: `tr`
* Ukrainian: `uk`
</div>

### `allow_public_update_access`
<div class="config-option">

Set to `1` to allow public access for `/update` (anybody can access and start the update job).
</div>

### `share`
<div class="config-option">

`share` defines which sharing buttons beneath the entry are visible. The following methods are supported:

<dl>
<dt><code>a</code></dt><dd><a href="https://developer.mozilla.org/en-US/docs/Web/API/Navigator/share">Web Share API</a>, when available</dd>
<dt><code>f</code></dt><dd>Facebook</dd>
<dt><code>t</code></dt><dd>Twitter</dd>
<dt><code>p</code></dt><dd>Pocket</dd>
<dt><code>d</code></dt><dd>Diaspora</dd>
<dt><code>m</code></dt><dd>Mastodon (requires <a href="#mastodon"><code>mastodon</code></a> option to be set, will only be available in selfoss 2.20 or later)</dd>
<dt><code>w</code></dt><dd>Wallabag (requires <a href="#wallabag"><code>wallabag</code></a> option to be set)</dd>
<dt><code>s</code></dt><dd>Wordpress (requires <a href="#wordpress"><code>wordpress</code></a> option to be set)</dd>
<dt><code>e</code></dt><dd>E-mail</dd>
<dt><code>c</code></dt><dd>Copy to clipboard</dd>
</dl>

Include the letters for methods you want to use. For example, if you would like to only show Facebook and Twitter share buttons, use `share=ft`.

Defaults to `share=atfpde`.
</div>

### `mastodon`
<div class="config-option">

URL of your Mastodon instance, for example `https://example.com`.
</div>

### `wallabag`
<div class="config-option">

URL of your [Wallabag](https://www.wallabag.org/) instance.
</div>

### `wallabag_version`
<div class="config-option">

Set to `2` or `1` depending on your wallabag version (`2` for version 2.x or `1` for version 1.x).
</div>

### `wordpress`
<div class="config-option">

URL of your WordPress blog for sharing links.
</div>

### `unread_order`
<div class="config-option">

Set to `asc` to read your unread items from the oldest to the newest, leave it empty or to `desc` to read from the newest to the oldest.
</div>

### `load_images_on_mobile`
<div class="config-option">

Set to `1` to allow image lazy loading on mobile devices.
</div>

### `auto_hide_read_on_mobile`
<div class="config-option">

Hide read articles on mobile devices.
</div>

### `scroll_to_article_header`
<div class="config-option">

Set to `0` to stop the interface from scrolling to the article header when an article is opened.
</div>

### `env_prefix`
<div class="config-option">

selfoss can use environment variables that start with this prefix as additional configuration options. This is useful for tools like Docker, where creating a config file is impractical. Defaults to `SELFOSS_`.
</div>

### `camo_domain`
<div class="config-option">

Camo domain used to proxy images (optional). See [atmos/camo](https://github.com/atmos/camo) for more details.
</div>

### `camo_key`
<div class="config-option">

Camo domain used to proxy images (optional). See [atmos/camo](https://github.com/atmos/camo) for more details.
</div>

### `show_thumbnails`
<div class="config-option">

If set to `0`, thumbnails are not shown in the collapsed view. Defaults to `1`.
</div>

### `datadir` {#datadir}
<div class="config-option">

Location of the data directory; especially useful when selfoss is installed to write-protected file system. `.htaccess` file (or equivalent configuration file for non-Apache web servers) will need to be adjusted accordingly.
</div>

### `reading_speed_wpm`
<div class="config-option">

Reading speed in words per minute used to calculate the estimated reading time of each article. On average, adults can read between 200 and 300 wpm and you can find many tools that can help you determine your reading speed on-line. If set to `0` (the default value), the estimated reading time is not shown.
</div>


docs/content/docs/administration/configuring.md
+++
title = "Configuring"
weight = 30
+++

selfoss can be configured using a configuration file or environment variables. All [configuration options](@/docs/administration/options.md) are optional.

## Using a file {#file}

This is the easiest way. Just create a `config.ini` file in the top-level selfoss directory and include any options you want to override.

The file uses [PHP’s variant](https://www.php.net/manual/en/function.parse-ini-file.php#refsect1-function.parse-ini-file-examples) of the [INI format](https://en.wikipedia.org/wiki/INI_file): each option is set on a separate line consisting of an option name, followed by `=` and an option value. Lines starting with `;` or empty lines will be ignored. If an option value contains any special characters such as `?{}|&~!()^"`, e.g. for a database password, it will need to be quoted like `db_password="life0fD4ng3r!"`.

For convenience, the release archive includes `config-example.ini` file containing the default configuration exported in INI format. To customize the settings, you can:

1. Rename `config-example.ini` to `config.ini`.
2. Edit `config.ini` and delete any lines you do not wish to override.

### Sample `config.ini` file which provides password protection

```ini
username=secretagent
password=$2y$10$xLurmBB0HJ60.sar1Z38r.ajtkruUIay7rwFRCvcaDl.1EU4epUH6
```

### Sample `config.ini` file with a MySQL database connection

```ini
db_type=mysql
db_host=localhost
db_database=selfoss
db_username=secretagent
db_password=life0fD4ng3r
```

### Location of config files {#location}

The configuration file path used will be in order:

1. The value of `SELFOSS_CONFIG_PATH` environment variable, if set, otherwise
2. `config.ini` in the directory from `SELFOSS_CONFIG_DIR` environment variable, if set, otherwise
3. `config.ini` in the selfoss root directory

If `SELFOSS_CONFIG_PATH` or `SELFOSS_CONFIG_DIR` environment variable is set, it must refer to a valid file or directory, respectively.

## Using environment variables {#env}

For environments such as Docker containers or <abbr title="Software-as-a-service">SaaS</abbr> inspired by [Twelve-Factor app](https://12factor.net/), where a persistent file system might not be available, you can also configure selfoss using environment variables.

The variables should start with a `SELFOSS_` prefix and be in all-caps. For example, if you want to set `auto_mark_as_read` option, use `SELFOSS_AUTO_MARK_AS_READ` variable.


NEWS.md
# selfoss news
## 2.20 – unreleased

**This version currently requires PHP 8.2 or newer.**

### New features
- YouTube spout now accepts handles (starting with `@` sign). ([#1412](https://github.com/fossar/selfoss/pull/1412))
- YouTube spout now includes the video description. ([#1412](https://github.com/fossar/selfoss/pull/1412))
- Mastodon share button was added. Can be enabled by adding `m` to `share` and setting `mastodon` pointing to your chosen instance. ([#1421](https://github.com/fossar/selfoss/pull/1421))
- Source filters can be negated, or limited to only title or only content. ([#1423](https://github.com/fossar/selfoss/pull/1423))
- Sources can be filtered based on item’s author, URL or categories. ([#1423](https://github.com/fossar/selfoss/pull/1423), [#1424](https://github.com/fossar/selfoss/pull/1424))
- Source filter expression is now validated whenever a source is modified. ([#1423](https://github.com/fossar/selfoss/pull/1423))
- RSS with enclosures spout now links audio enclosures (e.g. podcasts). ([#1574](https://github.com/fossar/selfoss/pull/1574))
- Garbage collection can be completely disabled by setting `items_lifetime=0`.
- Tamil (`ta`) translation was added.
- Configuration file path can be overridden with `SELFOSS_CONFIG_PATH` environment variable. Alternately, you can specify `SELFOSS_CONFIG_DIR` where selfoss will look for `config.ini`. ([#1603](https://github.com/fossar/selfoss/pull/1603))
- Password hash and database password can be now loaded from a file specified by `password_file` and `db_password_file` option, respectively. ([#1624](https://github.com/fossar/selfoss/pull/1624))

### Bug fixes
- Configuration parser was changed to *raw* method, which relaxes the requirement to quote option values containing special characters in `config.ini`. ([#1371](https://github.com/fossar/selfoss/issues/1371))
- Fix “Mark all as read” button not hiding marked articles in unread view, not updating the unread counts in menu properly, and not closing menu on mobile. ([#1388](https://github.com/fossar/selfoss/issues/1388)
- Re-added “Next” button on smartphones. ([#1406](https://github.com/fossar/selfoss/issues/1406)
- Fix compressed SVG (svgz) support. ([#1418](https://github.com/fossar/selfoss/pulls/1418)
- Fix article links containing HTML-special characters. ([#1407](https://github.com/fossar/selfoss/issues/1407))
- Reduce the chance of “Update all sources” button timing out. ([#1428](https://github.com/fossar/selfoss/pulls/1428), [#1430](https://github.com/fossar/selfoss/pulls/1430))
- Fix a log-in loop in client. ([#1429](https://github.com/fossar/selfoss/pulls/1429))
- Fix errors in Firefox’s private browsing mode.
- Fix exporting OPML when there are tags that look like numbers ([#1439](https://github.com/fossar/selfoss/pull/1439))
- Fix incorrect handling of tags in MySQL backend, which could result in OPML export being broken ([#1439](https://github.com/fossar/selfoss/pull/1439))
- Fix sharing to Wallabag 2. ([#1465](https://github.com/fossar/selfoss/pull/1465))
- Fix DB migration with SQLite that has double-quote string literals disabled (like on FreeBSD). ([#1489](https://github.com/fossar/selfoss/pull/1489))
- Look for favicon on correct address when feed URI has a port number. ([#1603](https://github.com/fossar/selfoss/pull/1603))

### Customization changes
- Custom spouts must explicitly pass `null` to `Item::__construct()` when they do not need the `extraData` argument. ([#1415](https://github.com/fossar/selfoss/pull/1415))
- Custom spout parameter declarations should now use constants from `Parameter` interface. ([#1409](https://github.com/fossar/selfoss/pull/1409))
- Custom spouts are expected to pass `HtmlString` object to items’ title and content. ([#1368](https://github.com/fossar/selfoss/pull/1368))
- Spouts can fetch item contents lazily by passing a function as `content` to `Item`. ([#1413](https://github.com/fossar/selfoss/pull/1413))
- Spouts’ `name`, `description` and `params` properties now require a type hint. ([#1425](https://github.com/fossar/selfoss/pull/1425))
- All classes other than spouts were made final. ([#1546](https://github.com/fossar/selfoss/pull/1546))
- The debugging level (in selfoss ≤ 2.18 set by modifying `src/common.php`) can be changed in the `config.ini` using `debug` key. ([#1298](https://github.com/fossar/selfoss/pull/1298))

### Other changes
- `tidy` PHP extension is now required if you want to use “Content extractor” spout. ([#1392](https://github.com/fossar/selfoss/pull/1392))
- Password hashing helper page will delegate the hashing to server again. ([#1401](https://github.com/fossar/selfoss/pull/1401))
- Content Extraction spout will no longer try to extract content we have already extracted. ([#1413](https://github.com/fossar/selfoss/pull/1413))
- Source filters are stricter, they need to start and end with a `/`. ([#1423](https://github.com/fossar/selfoss/pull/1423))
- OPML importer has been merged into the React client. ([#1442](https://github.com/fossar/selfoss/pull/1442))
- Web requests will send `Accept-Encoding` header. ([#1482](https://github.com/fossar/selfoss/pull/1482))
- Authentication system has been rewritten to allow more methods in the future. ([#1491](https://github.com/fossar/selfoss/pull/1491))
- Authentication will now also log user out when the credentials in the config change. ([#1491](https://github.com/fossar/selfoss/pull/1491))
- Requests from loopback IP address now give full access to all operations, not just update. Additionally, IPv6 loopback address is recognized and proxies are ignored. ([#1491](https://github.com/fossar/selfoss/pull/1491))
- [Tracy](https://tracy.nette.org/) is now used for error handling, resulting in much nicer error messages. ([#1298](https://github.com/fossar/selfoss/pull/1298))

#### For developers
- Back-end source code is now checked using [PHPStan](https://phpstan.org/). ([#1409](https://github.com/fossar/selfoss/pull/1409))

src/helpers/Configuration.php
<?php

declare(strict_types=1);

namespace Selfoss\helpers;

use Exception;
use ReflectionClass;
use ReflectionNamedType;
use Selfoss\helpers\Configuration\LoggerLevel;

/**
 * Configuration container.
 *
 * @see https://selfoss.aditu.de/docs/administration/options/ for more information about the configuration parameters.
 */
final class Configuration {
    /** @var string[] List of config values that should have variables interpolated. */
    public const INTERPOLATED_PROPERTIES = [
        'dbFile',
        'loggerDestination',
        'cache',
        'ftrssCustomDataDir',
    ];

    /** @var array<string, bool> Keeps track of options that have been changed. */
    private array $modifiedOptions = [];

    // Internal but overridable values.

    /** @internal */
    public string $datadir = __DIR__ . '/../../data';

    /** @internal */
    public string $cache = '%datadir%/cache';

    /** @internal */
    public string $ftrssCustomDataDir = '%datadir%/fulltextrss';

    // Rest of the values.

    /** Debugging level */
    public int $debug = 0;

    public string $dbType = 'sqlite';

    public string $dbFile = '%datadir%/sqlite/selfoss.db';

    public string $dbHost = 'localhost';

    public string $dbDatabase = 'selfoss';

    public string $dbUsername = 'root';

    private string $dbPasswordFile = '';

    public string $dbPassword = '';

    public ?int $dbPort = null;

    public ?string $dbSocket = null;

    public string $dbPrefix = '';

    public string $loggerDestination = 'file:%datadir%/logs/default.log';

    public LoggerLevel $loggerLevel = LoggerLevel::Error;

    public int $itemsPerpage = 50;

    public int $itemsLifetime = 30;

    public string $baseUrl = '';

    public string $username = '';

    private string $passwordFile = '';

    public string $password = '';

    public string $salt = 'lkjl1289';

    public bool $public = false;

    public string $htmlTitle = 'selfoss';

    public string $rssTitle = 'selfoss feed';

    public int $rssMaxItems = 300;

    public bool $rssMarkAsRead = false;

    public string $homepage = 'newest';

    public ?string $language = null;

    public bool $autoMarkAsRead = false;

    public bool $autoCollapse = false;

    public bool $autoStreamMore = true;

    public bool $openInBackgroundTab = false;

    public string $share = 'atfpde';

    public string $wallabag = '';

    public int $wallabagVersion = 2;

    public ?string $wordpress = null;

    public ?string $mastodon = null;

    public bool $allowPublicUpdateAccess = false;

    public string $unreadOrder = 'desc';

    public bool $loadImagesOnMobile = false;

    public bool $autoHideReadOnMobile = false;

    public string $envPrefix = 'selfoss_';

    public string $camoDomain = '';

    public string $camoKey = '';

    public bool $scrollToArticleHeader = true;

    public bool $showThumbnails = true;

    public int $readingSpeedWpm = 0;

    /**
     * @param array<string, string> $environment
     */
    public function __construct(?string $configPath = null, array $environment = []) {
        // read config.ini, if it exists
        if ($configPath !== null && file_exists($configPath)) {
            $config = parse_ini_file($configPath, false, INI_SCANNER_RAW);
            if ($config === false) {
                throw new Exception('Error loading config.ini');
            }
        } else {
            $config = [];
        }

        // overwrite config with ENV variables
        if (isset($config['env_prefix'])) {
            $this->envPrefix = $config['env_prefix'];
        }

        $reflection = new ReflectionClass(self::class);
        foreach ($reflection->getProperties() as $property) {
            $underscoreSeparatedName = preg_replace('([[:upper:]]+)', '_$0', $property->getName());
            assert($underscoreSeparatedName !== null, 'Regex must be valid');
            $configKey = strtolower($underscoreSeparatedName);

            if (isset($environment[strtoupper($this->envPrefix . $configKey)])) {
                // Prefer the value from environment variable if present.
                $value = $environment[strtoupper($this->envPrefix . $configKey)];
            } elseif (isset($environment[$this->envPrefix . $configKey])) {
                // Also try lowercase spelling.
                $value = $environment[$this->envPrefix . $configKey];
            } elseif (isset($config[$configKey])) {
                // Finally, try the value from config.ini.
                $value = $config[$configKey];
            } else {
                // Otherwise, just leave the default value.
                continue;
            }

            $value = trim($value);

            $nullable = false;
            $propertyType = null;

            $type = $property->getType();
            $nullable = $type !== null && $type->allowsNull();
            if ($type instanceof ReflectionNamedType) {
                $propertyType = $type->getName();
            }

            if ($nullable && $value === '') {
                // Keep the default value for empty nullables.
                continue;
            }

            $propertyName = $property->getName();
            if ($propertyType === 'bool') {
                $value = (bool) $value;
            } elseif ($propertyType === 'int') {
                $value = (int) $value;
            } elseif ($propertyType === 'string') {
                // Should already be a string.
            } elseif ($propertyType === LoggerLevel::class) {
                $value = LoggerLevel::tryFrom($value);
                if ($value === null) {
                    throw new Exception("Unsupported value “{$value}” for property “{$propertyName}”, must be one of " . implode(', ', array_map(fn(LoggerLevel $level) => $level->value, LoggerLevel::cases())) . '.', 1);
                }
            } else {
                throw new Exception("Unknown type “{$propertyType}” for property “{$propertyName}”.", 1);
            }

            $this->setValue($propertyName, $value);
        }

        // Interpolate variables in the config values.
        $datadir = $this->datadir;
        foreach (self::INTERPOLATED_PROPERTIES as $property) {
            $value = $this->{$property};
            $this->{$property} = str_replace('%datadir%', $datadir, $value);
        }

        $this->loadFromPath('password', 'password');
        $this->loadFromPath('dbPassword', 'db_password');
    }

    private function setValue(string $propertyName, mixed $value): void {
        $this->{$propertyName} = $value;
        $this->modifiedOptions[$propertyName] = true;
    }

    private function loadFromPath(string $propertyName, string $option): void {
        if ($this->isChanged($propertyName . 'File')) {
            if ($this->isChanged($propertyName)) {
                throw new Exception("You cannot set both `$option` and `{$option}_file` options.");
            }
            $value = @file_get_contents($this->{$propertyName . 'File'});
            if ($value !== false) {
                $this->setValue($propertyName, rtrim($value, "\r\n"));
            }
        }
    }

    /**
     * Checks whether given configuration option has been changed.
     */
    public function isChanged(string $key): bool {
        return isset($this->modifiedOptions[$key]);
    }
}
